### ISSUE TITLE
Questions regarding Section 3

### ISSUE THREAD

**Issue by nhattan417** [Datseris/Zero2Hero-JuliaWorkshop#13]
1. The variable `pars` in Notebook 3 (line 3, cell 102/138) is not defined. I think instead, we can directly use `testdata` inplace, or we have to change/define the variable `pars` before using it.
```julia
model = "linear"
testdata = @strdict a b c model
savename("prefix", pars, "jld2")
```

2. I think parameter for solution of "Bouncing ball" exercise should be `p₀ = [10, 0.99]` as mentioned in the instruction of Notebook 3 that use gravity constant $g=10$
https://github.com/Datseris/Zero2Hero-JuliaWorkshop/blob/3929cb090214ced54bd73f5eee0ecb806c8c667b/solutions_3.jl#L8

3. In the instruction of "Distribution quantile" exercise, I think we can write the formula (cell 133/138) as
$$\theta_{n+1} = \theta_n -\frac{\text{cdf}(\theta_n)−q}{\text{pdf}(\theta_n)}$$
instead of
$$\theta_{n+1} = \theta_n -\frac{\text{cdf}(\theta)−q}{\text{pdf}(\theta)}$$
In my case, at first glance I confused between $\theta$ and $\theta_n$ / $\theta_{n+1}$.

4. I have some errors using the solution for "DataFrame of chaos" exercise, heatmap plotting task. I get the message "StackOverflowError" when running the line
https://github.com/Datseris/Zero2Hero-JuliaWorkshop/blob/3929cb090214ced54bd73f5eee0ecb806c8c667b/solutions_3.jl#L98
When I delete 2 arguments `bs` and `cs` (the line becomes `fig, ax, hm = heatmap(heat)`), I do not get the error and get the heatmap successfully plotted. Can you re-run on your local to double check this?

5. In the solution for "DataFrame of chaos" exercise, you call directly Roessler built-in function
https://github.com/Datseris/Zero2Hero-JuliaWorkshop/blob/3929cb090214ced54bd73f5eee0ecb806c8c667b/solutions_3.jl#L60
While in my case I use
```julia
function RoesslerDynamicalSystem(u, p, t)
    x, y, z = u
    a, b, c = p
    xdot = - y - z
    ydot = x + a*y
    zdot = b + z*(x - c)
    return SVector(xdot, ydot, zdot)
end

ro = CoupledODEs(RoesslerDynamicalSystem, ones(3), [0.2, 0.2, 5.7])
```
The numbers `[0.2, 0.2, 5.7]` of parameters is what I have found at this [link of documentation](https://juliadynamics.github.io/PredefinedDynamicalSystems.jl/dev/#PredefinedDynamicalSystems.roessler). All the other things stay the same. However, I get slightly different plots

Built-in function | My own function
:-------------------------:|:-------------------------:
![Scatter plot when using built-in function](https://github.com/Datseris/Zero2Hero-JuliaWorkshop/assets/46997795/419ce194-eed0-4a23-a734-5e30a8cac877)  |  ![Scatter plot when using my own function](https://github.com/Datseris/Zero2Hero-JuliaWorkshop/assets/46997795/2ae1b8e2-43f1-4a19-b51d-1042b7a12c47)
![Heatmap when using built-in function](https://github.com/Datseris/Zero2Hero-JuliaWorkshop/assets/46997795/46f65393-76fb-46d2-a9d0-d0c2b286660c)  |  ![Heatmap when using my own function](https://github.com/Datseris/Zero2Hero-JuliaWorkshop/assets/46997795/e273539a-85b8-43a3-a6fc-b8b7087776df)

P/s: I do not have much knowledge in ODEs, DEs in general. So I will be very happy if you can correct me if I am wrong anywhere above. Furthermore, we currently do not have solution for "Textbook optimization problem" exercise, very happy to make a PR to contribute for that, and also fix some typos along the way.

**Comment by Datseris** [comment]
points 1-3 are all yes. point 5 is "the plots are the same". The miniscule differences you see is because of chaos.  For point 4, I don't encounter this problem. What version makie are you using?

Can you open a PR to fix the points 1-3? For point 5 let's leave it with the built-in function.

**Comment by nhattan417** [comment]
@Datseris 
My version of CairoMakie is v0.11.1
Yes, I have created a PR #14.

**Comment by Datseris** [comment]
i admit i don't understand why you would see the stack overflow for point 4. Can you create a MWE?

**Comment by nhattan417** [comment]
Maybe it is just because of my laptop. Sorry but what is mean by a MWE? 
I just copy and paste the whole same block of code (the solution for "DataFrame of chaos"). After running for ~30 seconds, I received the error message. I can upload the full error log if necessary.

**Comment by Datseris** [comment]
MWE stands for Minimal Working Example. A minimal piece of code that reproduces your error. As I said, I cannot reproduce it. For me, this runs perfectly fine:
```
julia> bs = 0.15:0.025:0.25
0.15:0.025:0.25

julia> cs = 4:0.1:6.0
4.0:0.1:6.0

julia> heat = rand(length(bs), length(cs));

julia> typeof(heat)
Matrix{Float64} (alias for Array{Float64, 2})

julia> fig, ax, hm = heatmap(bs, cs, heat)
```
and you quoted that the problem is at the `heatmap` function. I would recommend you to copy the solution code, and start removing more and more elements until you have the smallest code possible that reproduces your error. Clearly your error doesn't depend on simulating this Roessler systems, or on estimating lyapunov exponents, so you can remove all that stuff.

**Comment by nhattan417** [comment]
Hi, thanks for your clarification. I checked again the types and have the code here
```julia
using CairoMakie

bs = 0.15:0.025:0.25
cs = 4:0.1:6.0

heat = Matrix{Union{Float64, Missing}}(undef, length(bs), length(cs))
fill!(heat, 1)
fig, ax, hm = heatmap(bs, cs, heat)
```
However, if I change the variable `heat` to
```julia
heat = Matrix{Union{Float64}}(undef, length(bs), length(cs))
```
then everything is fine. I think the problem is something relating to type-unstable?
P/s: The Missing type coming from the step unstacking the DataFrame
https://github.com/Datseris/Zero2Hero-JuliaWorkshop/blob/ff0597a48fbd97a9d0c1611d67afe762959cca1c/solutions_3.jl#L96

**Comment by nhattan417** [comment]
I have found the solution for this. 

The problem comes from the function `DataFrames.unstack` which returns a DataFrame but in different type from the beginning. We initially have the datatypes are all `Float64` but after using `unstack`, the types become `Float64?` or `Union{Float64, Missing}` although we might not have missing values in the initial DataFrame. 
I double-checked the [examples in the documentation](https://dataframes.juliadata.org/stable/lib/functions/#DataFrames.unstack) which also have this behaviour.

I simply fix it by adding a new line right after
https://github.com/Datseris/Zero2Hero-JuliaWorkshop/blob/ff0597a48fbd97a9d0c1611d67afe762959cca1c/solutions_3.jl#L96
which then become
```julia
unstacked = unstack(heatdf, :b,  :H; renamecols = (x -> "H for b=$(x)"))
disallowmissing!(unstacked)
```
to change the type back to `Float64` and everything just works.

**Comment by Datseris** [comment]
Right, okay. Then, can you do a pull request that adds this extra line with an additional comment "Makie errors when trying to plot missing so we need to convert the type." 

In addition, I would strongly recommend opening an issue in the Makie repository. The error you should have been getting should be related to incorrect numeric type. Instead, you get a stack overflow error, which arrives from self-referential code.

**Comment by nhattan417** [comment]
@Datseris 
For your information, I have just found this completed issue MakieOrg/Makie.jl#3515 which is fixed in MakieOrg/Makie.jl#3597 and released on [Makie v0.20.7](https://github.com/MakieOrg/Makie.jl/releases/tag/v0.20.7).

I then tried update to newest version [Makie v0.21.3](https://github.com/MakieOrg/Makie.jl/releases/tag/v0.21.3) (CairoMakie v0.12.3) and your original code works just fine.
